ement-room.el

```diff
@@ -2910,63 +2910,46 @@ the first and last nodes in the buffer, respectively."
-                                  (not (read-marker-p data)))
-                                #'ewoc-prev)))
-            (when prev-node
-              ;; A previous node still exists: maybe we need to add a header.
-              (cl-typecase (ewoc-data prev-node)
-                (ement-event
-                 ;; Previous node is an event.
-                 (when (and (message-event-p (ewoc-data prev-node))
-                            (not (equal (ement-event-sender (ewoc-data prev-node))
-                                        (ement-event-sender (ewoc-data event-node)))))
-                   ;; Previous node is a message event with a different sender: insert
-                   ;; header.
-                   (insert-sender-before event-node)))
-                ((or ement-user ement-room-membership-events)
-                 ;; Previous node is a user or coalesced membership events: do not insert
-                 ;; header.
-                 nil)
-                (t
-                 ;; Previous node is not an event and not a read marker: insert header.
-                 (insert-sender-before event-node))))))
+                           (equal "m.room.message" (ement-event-type data)))))
+    ;; If we're not starting at a message node, scan back to the previous message. If none
+    ;; exist, scan forward to the next message. If we don't find any messages, there's
+    ;; nothing to do anyways.
+    (when (and start-node (not (message-event-p start-node)))
+      (setf start-node
+            (or (ement-room--ewoc-next-matching ewoc start-node #'message-event-p #'ewoc-prev)
+                (ement-room--ewoc-next-matching ewoc start-node #'message-event-p #'ewoc-next))))
+
+    ;; Set the end node to the next message after the end node, if any. This:
+    ;; 1. Makes the loop check simple (we continue going until we run out of nodes or our
+    ;;    we find the end node).
+    ;; 2. Ensures we fix up sender headers after any inserted messages.
+    (when end-node
+      (setf end-node (ement-room--ewoc-next-matching ewoc end-node #'message-event-p)))
+
+    (let* ((event-node start-node) prev-node)
+      (while (and event-node (not (eq event-node end-node)))
+        ;; Scan back for the previous message or user header, if any.
+        (setf prev-node (ement-room--ewoc-next-matching ewoc event-node
+                          (lambda (data)
+                            (or (ement-user-p data) (message-event-p data)))
+                          #'ewoc-prev))
+
+        (let ((sender (ement-event-sender (ewoc-data event-node))))
+          (cond
+           ;; No previous message/sender, insert a sender.
+           ((not prev-node) (ewoc-enter-before ewoc event-node sender))
+           ;; We have a previous sender, but it's wrong. Fix it.
+           ((ement-user-p (ewoc-data prev-node))
+            (unless (equal sender (ewoc-data prev-node))
+              (ewoc-set-data prev-node sender)))
```

Comment:
This can happen if we insert an event after a header but before another message. The read-marker fix below _should_ prevent this from happening (in most cases), but we might as well handle it.